coreneuron/nrnoc/multicore.cpp

```diff
@@ -68,77 +71,82 @@ the handling of v_structure_change as long as possible.
-                nt->end = 0;
+            for (i = 0; i < nrn_nthread; ++i) {
+                NrnThread& nt = nrn_threads[i];
+                nt._t = 0.;
+                nt._dt = -1e9;
+                nt.id = i;
+                nt._stop_stepping = 0;
+                nt.n_vecplay = 0;
+                nt._vecplay = nullptr;
+                nt.tml = nullptr;
+                nt._ml_list = nullptr;
+                nt.pntprocs = nullptr;
+                nt.presyns = nullptr;
+                nt.presyns_helper = nullptr;
+                nt.pnt2presyn_ix = nullptr;
+                nt.netcons = nullptr;
+                nt.weights = nullptr;
+                nt.n_pntproc = 0;
+                nt.n_presyn = 0;
+                nt.n_netcon = 0;
+                nt.n_weight = 0;
+                nt._ndata = 0;
+                nt._nidata = 0;
+                nt._nvdata = 0;
+                nt._data = nullptr;
+                nt._idata = nullptr;
+                nt._vdata = nullptr;
+                nt.ncell = 0;
+                nt.end = 0;
                 for (j = 0; j < BEFORE_AFTER_SIZE; ++j) {
-                    nt->tbl[j] = (NrnThreadBAList*)0;
+                    nt.tbl[j] = nullptr;
                 }
-                nt->_actual_rhs = 0;
-                nt->_actual_d = 0;
-                nt->_actual_a = 0;
-                nt->_actual_b = 0;
-                nt->_actual_v = 0;
-                nt->_actual_area = 0;
-                nt->_actual_diam = 0;
-                nt->_v_parent_index = 0;
-                nt->_permute = 0;
-                nt->_shadow_rhs = 0;
-                nt->_shadow_d = 0;
-                nt->_ecell_memb_list = 0;
-                nt->_sp13mat = 0;
-                nt->_ctime = 0.0;
-
-                nt->_net_send_buffer_size = 0;
-                nt->_net_send_buffer = (int*)0;
-                nt->_net_send_buffer_cnt = 0;
-                nt->_watch_types = NULL;
-                nt->mapping = NULL;
-                nt->trajec_requests = NULL;
-
-                nt->nrn_fast_imem = nullptr;
+                nt._actual_rhs = nullptr;
+                nt._actual_d = nullptr;
+                nt._actual_a = nullptr;
+                nt._actual_b = nullptr;
+                nt._actual_v = nullptr;
+                nt._actual_area = nullptr;
+                nt._actual_diam = nullptr;
+                nt._v_parent_index = nullptr;
+                nt._permute = nullptr;
+                nt._shadow_rhs = nullptr;
+                nt._shadow_d = nullptr;
+                nt._ecell_memb_list = nullptr;
+                nt._sp13mat = nullptr;
+                nt._ctime = 0.0;
+
+                nt._net_send_buffer_size = 0;
+                nt._net_send_buffer = nullptr;
+                nt._net_send_buffer_cnt = 0;
+                nt._watch_types = nullptr;
+                nt.mapping = nullptr;
+                nt.trajec_requests = nullptr;
+
+                nt.nrn_fast_imem = nullptr;
```

Comment:
replace this with class member initialiser.

Reply:
The reason why did not do this in the first place, is because it's not entirely clear at the moment when this function is called and if it can be called later again. Since it has side-effects it would not be safe to simply replace all this by class-member initialization. I prefer to do this after CoreNeuron is not a global object and it is clear what is accessed where and how.

Reply:
yes, but in this case it's straightforward because we allocate object and immediately initialise it (see two lines up): https://github.com/BlueBrain/CoreNeuron/blob/2b329dbbcea730906a8505753fa073418a5d861f/coreneuron/nrnoc/multicore.cpp#L90

Reply:
oops, yes you are right here - I mixed it up with initialization elsewhere.